File: exams/1617/translate/layout/student/java/VerticalLayout.java
import java.util.ArrayList;


public class VerticalLayout extends LinearLayout
{
    public VerticalLayout(ArrayList<Layout> children, double separation)
    {
        super(children, separation);
    }

    @Override
    public Box box()
    {
        Box result = new Box();

        result.height = Math.max(children.size() - 1, 0) * separation;

        for ( Layout child : children )
        {
            Box box = child.box();

            result.width += box.width;
            result.height = Math.max(box.height, result.height);
        }

        return result;
    }

    // Te vertalen als globale functie, d.i. een functie buiten de klasse
    public static Layout createHorizontalLayout(ArrayList<Layout> children, double separation)
    {
        return new HorizontalLayout(children, separation);
    }
}

File: exams/1617/translate/layout/solution/layout/vertical-layout.cpp
#include "vertical-layout.h"
#include <algorithm>


namespace
{
	class VerticalLayout : public LinearLayout
	{
	public:
		VerticalLayout(const std::vector<std::shared_ptr<Layout>>& children, double separation)
			: LinearLayout(children, separation) { }

		Box box() const
		{
			Box result;

			result.height = std::max((int)m_children.size() - 1, 0) * m_separation;

			for (auto& child : m_children)
			{
				auto box = child->box();

				result.width = std::max(box.width, result.width);
				result.height += box.height;
			}

			return result;
		}
	};
}

std::shared_ptr<Layout> create_vertical_layout(const std::vector<std::shared_ptr<Layout>>& children, double separation)
{
	return std::make_shared<VerticalLayout>(children, separation);
}
